libraries/platforms/dromajo-vcs/bsg_manycore_platform.cpp

```diff
@@ -0,0 +1,392 @@
+ * @return HB_MC_SUCCESS on success. Otherwise an error code defined in bsg_manycore_errno.h.
+ */
+int hb_mc_platform_get_config_at(hb_mc_manycore_t *mc, unsigned int idx, hb_mc_config_raw_t *config) {
+  hb_mc_platform_t *platform = reinterpret_cast<hb_mc_platform_t *>(mc->platform);
+  hb_mc_packet_t config_req_pkt, config_resp_pkt;
+  int num_entries = 0;
+  int err;
+
+  if (idx < HB_MC_CONFIG_MAX)
+  {
+    config_req_pkt.request.x_dst = (0 << 4) | 0;
+    config_req_pkt.request.y_dst = (1 << 3) | 0;
+    config_req_pkt.request.x_src = (0 << 4) | 0;
+    config_req_pkt.request.y_src = (1 << 3) | 1;
+    config_req_pkt.request.op_v2 = HB_MC_PACKET_OP_REMOTE_LOAD;
+    config_req_pkt.request.payload = 0;
+    config_req_pkt.request.reg_id = 0;
+    config_req_pkt.request.addr = MC_CONFIG_START_EPA_ADDR + idx;
+
+    // Note: Potentially dangerous to write to the FIFO without checking for credits
+    // We get back credits used and not credits remaining and without the configuration
+    // there is no way to know the credits remaining.
+    err = bp_hb_write_to_mc_bridge(&config_req_pkt);
+    if (err != HB_MC_SUCCESS) {
+      bsg_pr_err("Write to the host request FIFO failed!");
+      return err;
+    }
+
+    do {
+      err = bp_hb_get_fifo_entries(&num_entries, HB_MC_FIFO_RX_RSP);
+    } while (num_entries == 0 || err != HB_MC_SUCCESS);
+
+    err = bp_hb_read_from_mc_bridge(&config_resp_pkt, HB_MC_FIFO_RX_RSP);
+    if (err != HB_MC_SUCCESS) {
+      bsg_pr_err("Config read failed\n");
+      return HB_MC_FAIL;
+    }
+
+    uint32_t data = config_resp_pkt.response.data;
+    *config = (data != MC_HOST_OP_FINISH_CODE) ? data : 0;
+
+    return HB_MC_SUCCESS;
+  }
+
+  return HB_MC_INVALID;
+}
+
+/**
+ * Stall until the all requests (and responses) have reached their destination.
+ * @param[in]  mc     A manycore instance initialized with hb_mc_manycore_init()
+ * @param[in] timeout A timeout counter. Unused - set to -1 to wait forever.
+ * @return HB_MC_SUCCESS on success. Otherwise an error code defined in bsg_manycore_errno.h.
+ */
+int hb_mc_platform_fence(hb_mc_manycore_t *mc, long timeout) {
+  int credits_used, is_vacant, num_entries, err;
+  hb_mc_packet_t fence_req_pkt, fence_resp_pkt;
+  num_entries = 0;
+
+  hb_mc_platform_t *platform = reinterpret_cast<hb_mc_platform_t *>(mc->platform); 
+
+  if (timeout != -1) {
+    manycore_pr_err(mc, "%s: Only a timeout value of -1 is supported\n", __func__);
+    return HB_MC_NOIMPL;
+  }
+
+  // Prepare host packet to query TX vacancy
+  fence_req_pkt.request.x_dst = (0 << 4) | 0;
+  fence_req_pkt.request.y_dst = (1 << 3) | 0;
+  fence_req_pkt.request.x_src = (0 << 4) | 0;
+  fence_req_pkt.request.y_src = (1 << 3) | 1;
+  fence_req_pkt.request.op_v2 = HB_MC_PACKET_OP_REMOTE_LOAD;
+  fence_req_pkt.request.payload = 0;
+  fence_req_pkt.request.reg_id = 0;
+  fence_req_pkt.request.addr = MC_TX_VACANT_EPA_ADDR; // x86 Host address to poll tx vacancy
```

Comment:
Why is there a packet to query the TX vacancy

Reply:
In the whole system, the expectation is to have 3 FIFOs. These are modelled by the FIFOs sitting in the RTL testbench and DPI FIFOs allow C/C++ code to access them. In this Dromajo based system, there are 3 additional software FIFOs that are counterparts of what exists in the testbench that Dromajo uses to model the bridge as it sees it.

A fence is deemed to be complete when the FIFOs in the testbench are empty. Therefore, I send packets to the host to check TX vacancy using the API provided in the DPI class that @drichmond created. However, I discussed it with Dan and we concluded that it is sufficient to check the vacancy of the corresponding FIFO in Dromajo as long as the capacity of the FIFO in Dromajo is <= capacity of the FIFO in the testbench. I will be switching to this model in one of my next commits.

Reply:
> In the whole system, the expectation is to have 3 FIFOs. These are modelled by the FIFOs sitting in the RTL testbench and DPI FIFOs allow C/C++ code to access them. In this Dromajo based system, there are 3 additional software FIFOs that are counterparts of what exists in the testbench that Dromajo uses to model the bridge as it sees it.

So this is going away?

> 
> A fence is deemed to be complete when the FIFOs in the testbench are empty. Therefore, I send packets to the host to check TX vacancy using the API provided in the DPI class that @drichmond created. However, I discussed it with Dan and we concluded that it is sufficient to check the vacancy of the corresponding FIFO in Dromajo as long as the capacity of the FIFO in Dromajo is <= capacity of the FIFO in the testbench. I will be switching to this model in one of my next commits.

A fence is more than just TX empty, it is all credits have been returned. 

I think I'll understand this more when I see the README and understand the figure above